File: serene/src/Serene.Web/Modules/Documents/DocumentForm.cs
namespace Serene.Documents.Forms;

[FormScript("Documents.Document")]
[BasedOnRow(typeof(DocumentRow), CheckNames = true)]
public class DocumentForm
{
    public string Title { get; set; }

    [Updatable(false)]
    public DocumentType? DocumentType { get; set; }

    [DefaultValue("Draft"), ReadOnly(true)]
    public string State { get; set; }
}

File: serene/src/Serene.Web/Modules/ServerTypes/Documents/DocumentForm.ts
import { StringEditor, EnumEditor, PrefixedContext, initFormType } from "@serenity-is/corelib";
import { DocumentType } from "./DocumentType";

export interface DocumentForm {
    Title: StringEditor;
    DocumentType: EnumEditor;
    State: StringEditor;
}

export class DocumentForm extends PrefixedContext {
    static readonly formKey = 'Documents.Document';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!DocumentForm.init)  {
            DocumentForm.init = true;

            var w0 = StringEditor;
            var w1 = EnumEditor;

            initFormType(DocumentForm, [
                'Title', w0,
                'DocumentType', w1,
                'State', w0
            ]);
        }
    }
}

[DocumentType]; // referenced types